### PR TITLE
Feature - implementation of np.require in Cupy

### PR DESCRIPTION
**Pull request by niteya-shah**
I have added the implmentation and some tests and also added the documentation. Please review and suggest changes if any.

### CODE REVIEW

File: cupy/manipulation/kind.py [hunk 1]
```diff
@@ -24,4 +25,55 @@ def asfortranarray(a, dtype=None):
...
+    """
+
+    possible_flags = {'C': 'C', 'C_CONTIGUOUS': 'C', 'CONTIGUOUS': 'C',
+                      'F': 'F', 'F_CONTIGUOUS': 'F', 'FORTRAN': 'F',
+                      'O': 'OWNDATA', 'OWNDATA': 'OWNDATA'}
+
+    if not requirements:
+        return cupy.asanyarray(a, dtype=dtype)
+    else:
+        requirements = {possible_flags[x.upper()] for x in requirements}
```
Comment (emcastillo) [inline cupy/manipulation/kind.py]: Lets raise a value error with a descriptive message if the requirement is not in the flags keys.

File: tests/cupy_tests/manipulation_tests/test_kind.py [hunk 2]
```diff
@@ -58,3 +58,20 @@ def func(xp):
...
+            for parameter in flags:
+                assert arr.flags[parameter]
+                assert arr.dtype == dtype
+
+    @testing.for_all_dtypes()
+    def test_require_owndata(self, dtype):
+        x = cupy.zeros((2, 3, 4), dtype)
+        arr = x.view()
+        arr = cupy.require(arr, dtype, ['O'])
+        assert arr.flags['OWNDATA']
```
Comment (emcastillo) [inline tests/cupy_tests/manipulation_tests/test_kind.py]: Please add tests for invalid requirements, both c and f specified at the same time and other erroneous cases.

File: cupy/manipulation/kind.py [hunk 3]
```diff
@@ -24,4 +25,55 @@ def asfortranarray(a, dtype=None):
...
+    elif 'C' in requirements:
+        order = 'C_CONTIGUOUS'
+        requirements.remove('C')
+
+    arr = cupy.array(a, dtype=dtype, order=order, copy=False, subok=False)
+
+    for prop in requirements:
+        if not arr.flags[prop]:
+            arr = arr.copy(order)
+            break
```
Comment (emcastillo) [inline cupy/manipulation/kind.py]: If this is for `OWNDATA` check, shouldnt be easier to check it before 
```python
arr = cupy.array(a, dtype=dtype, order=order, copy=False, subok=False)
```
Such as
```python
copy = 'O' in requirements
arr = cupy.array(a, dtype=dtype, order=order, copy=copy, subok=False)
```

**Comment by niteya-shah** [comment]
@emcastillo I have made the changes. Can you please review them ?

File: tests/cupy_tests/manipulation_tests/test_kind.py (hunk 2 above)
Comment (emcastillo) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Sorry, can you add a test with empty requirements?

File: cupy/manipulation/kind.py [hunk 4]
```diff
@@ -24,4 +25,62 @@ def asfortranarray(a, dtype=None):
...
+    elif 'F' in requirements:
+        order = 'F_CONTIGUOUS'
+        requirements.remove('F')
+    elif 'C' in requirements:
+        order = 'C_CONTIGUOUS'
+        requirements.remove('C')
+
+    copy = 'OWNDATA' in requirements
+    try:
+        print("copy", copy)
```
Comment (emcastillo) [inline cupy/manipulation/kind.py]: remove print

File: tests/cupy_tests/manipulation_tests/test_kind.py (hunk 2 above)
Comment (niteya-shah) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Done

File: cupy/manipulation/kind.py (hunk 4 above)
Comment (niteya-shah) [reply cupy/manipulation/kind.py]: Sorry forgot this debugging message

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**